components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/impl/CacheBackedOrganizationManagementDAO.java

```diff
@@ -0,0 +1,358 @@
+import org.wso2.carbon.identity.organization.management.service.exception.OrganizationManagementServerException;
+import org.wso2.carbon.identity.organization.management.service.filter.ExpressionNode;
+import org.wso2.carbon.identity.organization.management.service.model.BasicOrganization;
+import org.wso2.carbon.identity.organization.management.service.model.Organization;
+import org.wso2.carbon.identity.organization.management.service.model.PatchOperation;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+
+import static org.wso2.carbon.identity.organization.management.service.constant.OrganizationManagementConstants.SUPER_ORG_ID;
+
+/**
+ * Cached DAO of organization management. All the DAO access should happen through this layer to ensure single point
+ * of caching.
+ */
+public class CacheBackedOrganizationManagementDAO implements OrganizationManagementDAO {
+
+    private final OrganizationManagementDAO organizationMgtDAO;
+
+
+    public CacheBackedOrganizationManagementDAO(OrganizationManagementDAO organizationMgtDAO) {
+
+        this.organizationMgtDAO = organizationMgtDAO;
+    }
+
+    @Override
+    public void addOrganization(Organization organization) throws OrganizationManagementServerException {
+
+        organizationMgtDAO.addOrganization(organization);
+    }
+
+    @Override
+    public boolean isOrganizationExistByName(String organizationName) throws OrganizationManagementServerException {
+
+        return organizationMgtDAO.isOrganizationExistByName(organizationName);
+    }
+
+    @Override
+    public boolean isOrganizationExistById(String organizationId) throws OrganizationManagementServerException {
+
+        return organizationMgtDAO.isOrganizationExistById(organizationId);
+    }
+
+    @Override
+    public String getOrganizationIdByName(String organizationName) throws OrganizationManagementServerException {
+
+        return organizationMgtDAO.getOrganizationIdByName(organizationName);
+    }
+
+    @Override
+    public Optional<String> getOrganizationNameById(String organizationId)
+            throws OrganizationManagementServerException {
+
+        String tenantDomain = resolveTenantDomain(organizationId);
+        if (tenantDomain == null) {
+            return organizationMgtDAO.getOrganizationNameById(organizationId);
+        }
+        OrganizationDetailsCacheEntry cachedOrgDetails = getOrganizationDetailsFromCache(organizationId, tenantDomain);
+        if (cachedOrgDetails != null && cachedOrgDetails.getOrgName() != null) {
+            return Optional.ofNullable(cachedOrgDetails.getOrgName());
+        }
+        Optional<String> orgName = organizationMgtDAO.getOrganizationNameById(organizationId);
+        if (cachedOrgDetails != null) {
+            cachedOrgDetails.setOrgName(orgName.get());
+        } else {
+            addOrganizationDetailsToCache(organizationId, orgName.get(), tenantDomain);
```

Comment:
apply the same comment https://github.com/wso2-extensions/identity-organization-management-core/pull/51/files#r1112276186